### ISSUE TITLE
use SegmentNumber from the segment metadata exporting DICOM SEG.

### ISSUE THREAD

**Issue by IbrahimCSAE** [dcmjs-org/dcmjs#339]
Hello,

I would like to ask if it's possible to use the SegmentNumber from the cornerstoneTools segmentation metadata when exporting DICOM SEG, rather than being limited to sequential segment numbers. Currently, if I have non-sequential segments (e.g. 3 and 4), when I export the segmentation using the dcmjs adapter, it will assign sequential segment numbers (1 and 2) when exported.

1 - Psoas Muscle -> exported index will map to this segment
2 - Subcutaneous Fat    -> exported index will map to this segment
3 - Visceral Fat -> user drew this segment
4 -  Skeletal Wall Muscle -> user drew this segment

![Screenshot_2](https://user-images.githubusercontent.com/93064150/220897744-83c3cd71-e760-48f1-910d-94b4af6dbbfd.png)

This change would be useful for our web app as we have hard-coded behavior based on the SegmentNumber that spans multiple areas of the app, including activating the correct segment for drawing, setting the correct color for the segment, the correct name in the UI, threshold values for the brush for that segment, configuring the cornerstone colorLUT, labelmap3Ds, labelmap2Ds, etc. We also have a fixed set of segments, which is why we have taken a hard-coded approach. Our backend also has AI models that follow these segment numbers strictly when it generates SEGs, making it easy for the frontend to identify what should be displayed in the UI.

Currently, if a user exports their drawn segmentation to the cloud and the indexes for non-sequential segments get turned into sequential segments, most of the logic that depends on the segment number will fail **(next time they retrieve it from the cloud, all the mapping will be wrong)**. This applies whether they draw manually or use any of our AI models and then overwrite the SEG when using dcmjs. This is also problematic when a user ovewrites a pre existing SEG in the cloud, when we regenerate the SEG with dcmjs and send it back to S3 or some other storage, that updated segmentation will have wrong index's, so it affects **new generations** or **overwriting existing** SEGs in the cloud storage.

I understand that this approach of linking behavior to the SegmentNumber might not be reliable and I should look into better approaches (maybe ID the segments some other way? Using a database and just ignoring the SegmentNumber in the meta entirely?). However, I would still like to ask if it is possible to have exports with the SegmentNumber. I have been using a fork of the library with the changes I mentioned, and it has been working okay for our needs.

```
 SegmentNumber: (SegmentSequence.length + 1).toString()
```
```
 SegmentNumber: Segment.SegmentNumber
```

https://github.com/dcmjs-org/dcmjs/blob/b4732d7b677b8cbb7ae12276019e18371a4864a6/src/derivations/Segmentation.js#L408

I acknowledge that this approach could fail if developers doesn't set the metadata for the segment in cornerstoneTools before exporting the SEG, or if they don't add the SegmentNumber in generated metadata. One way to decide whether to use sequential numbers or the metadata for the segments could be to check if all the segments contain the SegmentNumber field, then we proceed with the SegmentNumbers, otherwise fallback to sequential?

```
segmentsLength = segments.length
segmentNumbersLenght = segments.filter(segment => segment.SegmentNumber !== undefined)
if (segmentNumbersLength == segmentsLength){
 SegmentNumber: Segment.SegmentNumber
}else{
 SegmentNumber: (SegmentSequence.length + 1).toString()
}
```

In summary, can we use the SegmentNumber from the user-generated metadata when exporting the SEG? 

Any help is really appreciated, thank you, this library is great and It does a lot of amazing things. Thanks for all the efforts put into this!

**Comment by fedorov** [comment]
The standard is very clear in requiring `SegmentNumber` to be assigned sequentially: https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.8.20.2.4.html#sect_C.8.20.2.4

> Segment Number (0062,0004) shall be unique within each Instance, start at a value of 1, and increase monotonically by 1.

**Comment by fedorov** [comment]
Adding @dclunie, as I think it will be good for him to be aware of the issue you are trying to address.

**Comment by IbrahimCSAE** [comment]
> The standard is very clear in requiring `SegmentNumber` to be assigned sequentially: https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.8.20.2.4.html#sect_C.8.20.2.4
> 
> > Segment Number (0062,0004) shall be unique within each Instance, start at a value of 1, and increase monotonically by 1.

Thanks for sharing, It looks like I should rethink my entire approach to the problem then. I was not aware that it has to be sequential.

**Comment by pieper** [comment]
@IbrahimCSAE Thanks for your careful attention to this issue - it's a very important application level concept.  It is true that many systems rely on numbering conventions to identify segments, but independent of dicom, this can lead to issues like you describe when, for example, a new segment type is introduced or a previous segment gets subdivided.  There's always some versioned "map" that goes from these numbers to some other concept like anatomy, color, etc.

The dicom approach is to use other codes like snomed so you can describe the intent or meaning of the segment.  These are the 'coding schemes' and 'coded concepts' you'll see in the standard and there are other mechanisms to do things like provide an id to a lesion so you can track it over time.  If this is too complex or none of the standard schemes fit your application, you can define a custom scheme, with a designator like "99MYAPP" and put your segment numbers there so that your round trips to dicom seg are lossless.

**Comment by IbrahimCSAE** [comment]
@pieper 

Thank you for explaining the benefits of using coding schemes and coded concepts to identify segments in DICOM.

I took a look at the standard and this will definitely fit my use case. I appreciate it!

**Comment by fedorov** [comment]
Another "feature" of the standard you should be aware of is the `TrackingID`/`TrackingUID` attributes that you can assign at the segment level. Using those you can link segments corresponding to the same structure segmented across series/modalities/timepoints: https://dicom.innolitics.com/ciods/segmentation/segmentation-image/00620002/00620021.

**Comment by gacou54** [comment]
Hello, thank you for the excellent description of the problem @IbrahimCSAE .

We encountered the same issue. We are working on a modified version of OHIF where annotators answer predetermined questions (characteristics) about each segment. The characteristics are then written to a DICOM SR (similar idea to https://github.com/QIICR/QuantitativeReporting).

We were using the `SegmentIndex` of the Segment objects (in OHIF) to link the DICOM SR characteristics to the DICOM SEG segment.

We noticed the problem after producing 900+ annotations (the issue occurs when annotators delete segments in OHIF). For example, if a user had three segments in OHIF:
```javascript
{
  segmentIndex1: { segmentIndex: 1, ...segmentsdata},
  segmentIndex2: { segmentIndex: 2, ...segmentsdata},
  segmentIndex3: { segmentIndex: 3, ...segmentsdata},
}
```
 our SR object would also have: 
```javascript
{
  segmentIndex1: { segmentIndex: 1, ...characteristicsData},
  segmentIndex2: { segmentIndex: 2, ...characteristicsData},
  segmentIndex3: { segmentIndex: 3, ...characteristicsData},
}
```

Upon deleting the second segment, we are left with 
- Segments: `{1: seg1, 3: seg3}` 
- Characteristics `{1: char1, 3: char3}`. 

However, after saving as DICOM, the `SegmentNumber`s of the SEG become **1 and 2**, whereas we had assumed they would remain **1 and 3**. The `ReferencedSegmentNumber`s in our DICOM SR were **1 and 3**. This leads to cases where elements in the `ContentSequence` in the DICOM SR refers to an incorrect `ReferencedSegmentNumber`.

I simply want to raise awareness with this comment for people that might encounter this behavior. To my understaning, we shouldn't rely on `SegmentIndex`/`SegmentNumber` for linking derived DICOM objects (like SR)   when writing a DICOM SEG via dcmjs/Cornerstone.js/OHIF.

**Comment by fedorov** [comment]
> To my understaning, we shouldn't rely on SegmentIndex/SegmentNumber for linking derived DICOM objects (like SR) when writing a DICOM SEG via dcmjs/Cornerstone.js/OHIF.

In DICOM, you are not supposed to modify an instance. If you follow the standard, you would make a new DICOM SEG after deleting a segment, and create a new DICOM SR referencing it. You could reference prior instances from the new ones to establish provenance. This may not be practical, but modifying existing DICOM in-place is always risky and you should expect problems if you do that.

**Comment by gacou54** [comment]
I understand about the DICOM standard regarding instance modification.

My comment isn't about modifying an existing DICOM SEG file.
We're referring to the in-memory segment representation in OHIF before writing to DICOM SEG. We had assumed that SegmentIndex/SegmentNumber in OHIF’s segment objects would correspond directly to the SegmentNumber in the final DICOM SEG file, but that’s not guaranteed.

Specifically, when saving to DICOM SEG, the SegmentNumber is assigned as:
```js
SegmentNumber: (SegmentSequence.length + 1).toString()
```
rather than preserving:
```js
SegmentNumber: Segment.SegmentNumber
```
This renumbering can break references between the SEG and any derived SR created before the DICOM export if we are not aware of this behavior.

I recognize that this behavior aligns with the DICOM standard. Once the SEG is written, its `SegmentNumber`s can reliably be referenced. The lesson for us (and perhaps others) is to avoid relying on OHIF’s in-memory segment indices when constructing references in derivative objects like SR.

**Comment by pieper** [comment]
As I recall the SEG objects are required to be sequentially numbered starting at 1, so dcmjs is just following that.

You probably should assign a tracking UID to the segment, and then the SR can reference that.

https://dicom.innolitics.com/ciods/segmentation/segmentation-image/00620002/00620021

Here's what I'm referring to in case the deep link doesn't work:

<img width="1503" height="786" alt="Image" src="https://github.com/user-attachments/assets/7df8a064-b0ad-42bb-aeb6-8a550ce605f5" />

**Comment by pieper** [comment]
(ha ha, I see now @fedorov already posted this link above!)

**Comment by fedorov** [comment]
@gacou54 thank you for the clarification, I misunderstood your earlier comment. I don't know if the existing behavior is intended or not.

@pieper yes, you can assign that UID, but measurements in SR that are derived from a segment must always include SegmentNumber to reference it - this is what is used by IMAGE content item image reference, I think this is the right link: https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.4.html.

I would think that for in-memory management, either dcmjs could provide convenience functions to clean up references, or the user would need to take care of this in a situation where a segment is deleted.

**Comment by gacou54** [comment]
No worries, I am learning a lot today :). Indeed maybe it's better to use the Tracking UID. However using the `SegmentNumber` felt natural when referencing a Segment from a the DICOM SR using this, hence why we did go that route. https://dicom.innolitics.com/ciods/basic-text-sr/sr-document-content/00081199/0062000b

Our solution will likely to fix our in-memory Characteristics before writing the SEG and SR (basically we keep the same order in the segments and our characteristics objects, meaning we will rewrite the segmentIndex of those objects in OHIF).

**Comment by pieper** [comment]
> I would think that for in-memory management, either dcmjs could provide convenience functions to clean up references, or the user would need to take care of this in a situation where a segment is deleted.

In the "dcmjs -- cornerstone3d -- OHIF" path I think it should be the higher level adapter code to handle this kind of case.
